SSA/Projects/Scf/Scf.lean

```diff
@@ -51,52 +57,138 @@ instance : OpSignature Op Ty where
+
+namespace LoopBody
+
 /-- Convert a function `f` which is a single loop iteration into a function
   that iterates and updates the loop counter. -/
-def loop_counter_decorator (δ: Int) (f : Int → α → α) : Int × α → Int × α :=
+def CounterDecorator (δ: Int) (f : LoopBody α) : Int × α → Int × α :=
   fun (i, v) => (i + δ, f i v)
 
-/-- evaluating a function that does not access the index (const_index_fn) -/
-theorem loop_counter_decorator_const_index_fn_eval
-  (δ : Int) (i : Int) (vstart : α) (f : Int → α → α) (f' : α → α) (hf : f = fun i a => f' a) :
-  (loop_counter_decorator δ f) (i, vstart) = (i + δ, f' vstart) := by
-  simp[loop_counter_decorator, hf]
 
+/-- A loop body is invariant if it does not depend on the index. -/
+def IndexInvariant (f : LoopBody t) : Prop :=
+  ∀ (i j : Int) (v : t), f i v = f j v
 
-/-- iterating a function that does not access the index (const_index_fn) -/
-theorem loop_counter_decorator_const_index_fn_iterate
-  (δ : Int) (i : Int) (vstart : α) (f : Int → α → α) (f' : α → α) (hf : f = fun i a => f' a) (k : ℕ) :
-  (loop_counter_decorator δ f)^[k] (i, vstart) = (i + k * δ, f'^[k] vstart) := by
-  induction k generalizing i vstart
-  case zero =>
-    simp
-  case succ n ihn =>
-    simp
-    simp[loop_counter_decorator] at ihn ⊢
-    simp[ihn]
-    simp[hf]
-    linarith
+/-- Evaluating a loop index invariant function is the same as evaluating it at 0 -/
+theorem eval (f : LoopBody t) (hf : LoopBody.IndexInvariant f) (i : Int) (v : t) :
+  f i v = f 0 v := by unfold LoopBody.IndexInvariant at hf; rw [hf]
+
+/-- Loop invariant functions can be simulated by a simpler function -/
+def atZero (f : LoopBody t) : t → t := fun v => f 0 v
+
+/-- If there exists a function `g : t → t` which agrees with `f`, then `f` is loop index invariant. -/
+theorem eq_invariant_fn
+    (f : LoopBody t) (g : t → t) (hf : ∀ (i : Int) (v : t), f i v = g v) :
+    LoopBody.IndexInvariant f /\ atZero f = g:= by
+      apply And.intro
+      case left =>
+        intros i j v
+        rw [hf i v, hf j v]
+      case right =>
+        simp [atZero]
+        funext v
+        rw [hf 0 v]
+
+end LoopBody
+
+namespace LoopBody.IndexInvariant
 
+/-- Evaluating a loop index invariant function is the same as evaluating it at Invariant -/
+@[simp]
+theorem eval' {f : LoopBody t} (hf : LoopBody.IndexInvariant f) (i : Int) (v : t) :
+  f i v = f.atZero v := by unfold LoopBody.IndexInvariant at hf; rw [LoopBody.atZero, hf]
+
+/-- iterating a loop invariant function gives a well understood answer: the iterates of the function. -/
+@[simp]
+theorem iterate {f : LoopBody t} (hf : LoopBody.IndexInvariant f) (δ : Int)  (i₀ : Int) (v₀ : t) (k : ℕ) :
+  (f.CounterDecorator δ)^[k] (i₀, v₀) = (i₀ + δ * k, (f.atZero)^[k] v₀) := by
+    induction k generalizing i₀ v₀
+    case zero => simp
+    case succ i hi =>
+      simp
+      rw [hi]
+      simp [LoopBody.CounterDecorator]
+      simp [eval' hf]
+      linarith
```

Comment:
We indent the type with 4 spaces, and the body with 2 (at least, this is the Mathlib style). 

You should check this across the whole file, there's more defs that break this convention.

Reply:
Fixed across the file.